flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraph.java

```diff
@@ -1485,6 +1525,20 @@ else if (!isRestartable && transitionState(currentState, JobStatus.FAILED, failu
 		}
 	}
 
+	public void failJob(Throwable cause) {
+		if (state == JobStatus.FAILING || state.isGloballyTerminalState()) {
+			return;
+		}
+
+		transitionState(JobStatus.FAILING);
+		initFailureCause(cause);
+
+		cancelVerticesAsync().whenComplete((aVoid, throwable) -> {
+			transitionState(JobStatus.FAILED);
```

Comment:
How about using transitionState(state, JobStatus.FAILED, cause)?

Reply:
I think that's better – same for when we transition to `FAILING`, right?

Reply:
The transition log will contain the cause if there is any.
Without the cause seems we do not log the information that why the job is FAILING/FAILED?

Reply:
I originally meant that your proposal is better/makes sense. I was talking about line 1533 where we transition to `FAILING`. Do you think we should log the exception there as well?

Reply:
Oh I see. I think we can do that for transition to `FAILING` as well.

Reply:
Fixed